lexicon/providers/valuedomain.py

```diff
@@ -0,0 +1,227 @@
+
+    def __init__(self, config):
+        super(Provider, self).__init__(config)
+        self.nameserver = "valuedomain1"
+        self.ttl = 3600
+        self.domain_id = None
+        self.api_endpoint = "https://api.value-domain.com/v1"
+
+    def _authenticate(self):
+        payload = self._get("/domains/{0}/dns".format(self.domain))
+        self.ttl = int(payload["results"]["ttl"])
+        self.nameserver = payload["results"]["ns_type"]
+
+        if payload["results"]["domainname"] == self.domain:
+            self.domain_id = payload["results"]["domainid"]
+            return
+
+        raise Exception("No domain found")
+
+    # Create record. If record already exists with the same content, do nothing'
+    def _create_record(self, rtype, name, content):
+        name = self._relative_name(name)
+        resource_record_sets = self._get_resource_record_sets()
+        index = self._find_resource_record_set(
+            resource_record_sets, identifier=None, rtype=rtype, name=name, content=content
+        )
+        if index >= 0:
+            LOGGER.debug("record already exists")
+            return True
+
+        resource_record_sets.append((rtype.lower(), name, self._bind_format_target(rtype, content)))
+        self._update_resource_record_sets(resource_record_sets)
+        LOGGER.debug("create_record: %s", True)
+        return True
+
+    @staticmethod
+    def _identifier(record):
+        sha256 = hashlib.sha256()
+        sha256.update(('type=' + record[0].lower()).encode('utf-8'))
+        sha256.update(('name=' + record[1]).encode('utf-8'))
+        sha256.update(('data=' + record[2]).encode('utf-8'))
+        return sha256.hexdigest()[0:7]
+
+    # List all records. Return an empty list if no records found
+    # type, name and content are used to filter records.
+    # If possible filter during the query, otherwise filter after response is received.
+    def _list_records(self, rtype=None, name=None, content=None):
+        records = []
+        full_name = None
+        if name:
+            full_name = self._full_name(name)
+
+        for record in self._get_resource_record_sets():
+            processed_record = {
+                "type": record[0].upper(),
+                "name": self._full_name(record[1]),
+                "ttl": self.ttl,
+                "content": record[2],
+                'id': Provider._identifier(record),
+            }
+            if rtype and processed_record["type"].lower() != rtype.lower():
+                continue
+            if full_name and processed_record["name"] != full_name:
+                continue
+            if content and processed_record["content"] != content:
+                continue
+            records.append(processed_record)
+
+        LOGGER.debug("list_records: %s", records)
+        return records
+
+    # Create or update a record.
+    def _update_record(self, identifier=None, rtype=None, name=None, content=None):
```

Comment:
You can make the function a little more flexible regarding the inputs provided.

First, as for delete action, we can check for identifier, and update the record, or try the other parameters to find the matching one(s).

Second, if an identifier is not provided, you can require only rtype and name to be set (not content). In this case, you need to handle the situation where two record would be retrieved for a given rtype+name (quite common with TXT entries):
* either you warn the user and take the first
* or you error out indicating to the user that more than one potential record to update was found.